### ISSUE TITLE
Document that hypothesis is thread-unsafe

### ISSUE THREAD

**Issue by ngoldbaum** [Quansight-Labs/free-threaded-compatibility#116]
See https://hypothesis.readthedocs.io/en/latest/details.html#thread-safety-policy and https://github.com/indygreg/python-zstandard/pull/243 for a time this came up in the real world.

**Comment by ngoldbaum** [comment]
@zac-hd as a hypothesis maintainer I'd appreciate your take on my assessment here.

**Comment by Zac-HD** [comment]
You're correct that Hypothesis is not threadsafe.  I haven't yet worked out what to do about that given free-threading changes; it'd be great to support but also we really do have to manage some global state.

Upstream issue: https://github.com/HypothesisWorks/hypothesis/issues/4028

**Comment by ngoldbaum** [comment]
It's the same as pytest, it's fine to spawn worker threads in a hypothesis test but hypothesis itself has a lot of thread safety issues so if you use constructs provided by hypothesis in a multithreaded context or use hypothesis itself in many threads than you will likely hit issues. Lots of race conditions around global registries and that sort of thing. There's also an on-disk results db that may or may not have thread safety issues.

**Comment by Zac-HD** [comment]
The database is safe for concurrent use from multiple threads (it's a very basic best-effort KV store), entirely correct on the rest.

**Comment by ngoldbaum** [comment]
This is done now: https://py-free-threading.github.io/testing/#hypothesis-is-not-thread-safe
